### PR TITLE
Spark 1.2.1 upgrade

### PR DESCRIPTION
**Pull request by hamnis**
o Bumped some versions to get ready for Scala 2.11.
o Changed some references to include scala.major so we can upgrade easier
o Removed a duplicated implicit so we dont get errors in 2.11
o Fixed cleanup of tests.

### CODE REVIEW

File: math-scala/src/main/scala/org/apache/mahout/math/drm/package.scala [hunk 1]
```diff
@@ -80,9 +80,6 @@ package object drm {
   /** Just throw all engine operations into context as well. */
   implicit def ctx2engine(ctx: DistributedContext): DistributedEngine = ctx.engine
 
-  implicit def drm2drmCpOps[K: ClassTag](drm: CheckpointedDrm[K]): CheckpointedOps[K] =
```
Comment (dlyubimov) [inline math-scala/src/main/scala/org/apache/mahout/math/drm/package.scala]: why is this removed?

File: spark/src/test/scala/org/apache/mahout/sparkbindings/test/DistributedSparkSuite.scala [hunk 2]
```diff
@@ -58,6 +58,11 @@ trait DistributedSparkSuite extends DistributedMahoutSuite with LoggerConfigurat
   }
 
 
+  override protected def afterAll(configMap: ConfigMap): Unit = {
```
Comment (dlyubimov) [inline spark/src/test/scala/org/apache/mahout/sparkbindings/test/DistributedSparkSuite.scala]: oh. this is why we kept growing unused contexts in tests? 

if so, good catch. thank you.

Comment (hamnis) [inline spark/src/test/scala/org/apache/mahout/sparkbindings/test/DistributedSparkSuite.scala]: yes.

File: math-scala/src/main/scala/org/apache/mahout/math/drm/package.scala (hunk 1 above)
Comment (hamnis) [inline math-scala/src/main/scala/org/apache/mahout/math/drm/package.scala]: it already exits here: https://github.com/apache/mahout/blob/master/math-scala/src/main/scala/org/apache/mahout/math/drm/RLikeDrmOps.scala#L137. if you upgrade to 2.11 this would crash. and both are imported into scope. Why this does not crash on 2.10 i have no idea.
Not sure which one that should be removed, but I chose this one.

Comment (dlyubimov) [inline math-scala/src/main/scala/org/apache/mahout/math/drm/package.scala]: Hm. indeed. ok.


Normally we import both and i'd say drm._ is the place for it. RLikeDrmOps
deals with R dialect specifically so i'd say if i had to remove it, i'd
remove it from RLikeDrmOps (assuming all tests pass).

Comment (dlyubimov) [inline math-scala/src/main/scala/org/apache/mahout/math/drm/package.scala]: BTW Erlend thank you for looking into this. Much appreciated!

**Comment by dlyubimov** [comment]
ok i'd say if tests are passing this looks good to me with exception of switching the removed code snippet as discussed. +1 with contingency.

**Comment by hamnis** [comment]
should i just change this back as a new commit, or do I rebase the commit as discussed?

**Comment by dlyubimov** [comment]
You can just add another commit to this PR (i.e. to your source github
branch). You can string as many commits as you want in a single PR. as long
as they merge to master without conflicts. we will squash the history to a
single commit in the end. You may find information about this process here
: http://mahout.apache.org/developers/github.html

Once the branch is committed to master, you can't use it as a base for new
commits -- you'd have to fork a new one off the master (because squashed
history would be inconsistent with master's).

**Comment by hamnis** [comment]
Done. All tests pass. please check that it works on someone elses machine.

**Comment by dlyubimov** [comment]
thank you. 

we normally let stuff sit for 2-3 days to see if anyone is going to flag stuff, esp. since this is changing spark version, I want to give it a week or so so if nobody comes forward i will push in on 2/20.

**Comment by hamnis** [comment]
Cool =)

**Comment by dlyubimov** [comment]
module h20 still does not compile for me. 
On top of it, Pat's stuff is failing in tests ... not sure why. must be me. 

2.168.11.4:0, shutting down Netty transport
855 [sparkDriver-akka.actor.default-dispatcher-6] ERROR akka.remote.transport.netty.NettyTransport  - failed to bind to Intel-KUBU/192.168.11.4:0, shutting down Netty transport
*** RUN ABORTED ***
  java.net.BindException: Failed to bind to: Intel-KUBU/192.168.11.4:0: Service 'sparkDriver' failed after 16 retries!
  at org.jboss.netty.bootstrap.ServerBootstrap.bind(ServerBootstrap.java:272)
  at akka.remote.transport.netty.NettyTransport$$anonfun$listen$1.apply(NettyTransport.scala:393)
  at akka.remote.transport.netty.NettyTransport$$anonfun$listen$1.apply(NettyTransport.scala:389)
  at scala.util.Success$$anonfun$map$1.apply(Try.scala:206)
  at scala.util.Try$.apply(Try.scala:161)
  at scala.util.Success.map(Try.scala:206)
  at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
  at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
  at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
  at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.processBatch$1(BatchingExecutor.scala:67)
  ...

**Comment by dlyubimov** [comment]
i probably will push it anyway, see what automatic build has to say.

**Comment by andrewpalumbo** [comment]
could you post the h2o errors that you're getting?

**Comment by dlyubimov** [comment]
NFO] Compiling 28 source files to
/home/dmitriy/projects/github/mahout-commits/h2o/target/classes at
1424466468753
[ERROR] error: error while loading <root>, error in opening zip file
[ERROR] error: scala.reflect.internal.MissingRequirementError: object
scala.runtime in compiler mirror not found.
[ERROR]         at
scala.reflect.internal.MissingRequirementError$.signal(MissingRequirementError.scala:16)
[ERROR]         at
scala.reflect.internal.MissingRequirementError$.notFound(MissingRequirementError.scala:17)
[INFO]  at
scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:48)
[INFO]  at
scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:40)
[INFO]  at
scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:61)
[INFO]  at
scala.reflect.internal.Mirrors$RootsBase.getPackage(Mirrors.scala:172)
[INFO]  at
scala.reflect.internal.Mirrors$RootsBase.getRequiredPackage(Mirrors.scala:175)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.RuntimePackage$lzycompute(Definitions.scala:183)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.RuntimePackage(Definitions.scala:183)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.RuntimePackageClass$lzycompute(Definitions.scala:184)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.RuntimePackageClass(Definitions.scala:184)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.AnnotationDefaultAttr$lzycompute(Definitions.scala:1024)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.AnnotationDefaultAttr(Definitions.scala:1023)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.syntheticCoreClasses$lzycompute(Definitions.scala:1153)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.syntheticCoreClasses(Definitions.scala:1152)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.symbolsNotPresentInBytecode$lzycompute(Definitions.scala:1196)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.symbolsNotPresentInBytecode(Definitions.scala:1196)
[INFO]  at
scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1261)
[INFO]  at scala.tools.nsc.Global$Run.<init>(Global.scala:1290)
[INFO]  at scala.tools.nsc.Driver.doCompile(Driver.scala:32)
[INFO]  at scala.tools.nsc.Main$.doCompile(Main.scala:79)
[INFO]  at scala.tools.nsc.Driver.process(Driver.scala:54)
[INFO]  at scala.tools.nsc.Driver.main(Driver.scala:67)
[INFO]  at scala.tools.nsc.Main.main(Main.scala)
[INFO]  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[INFO]  at
sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
[INFO]  at
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[INFO]  at java.lang.reflect.Method.invoke(Method.java:606)
[INFO]  at
org_scala_tools_maven_executions.MainHelper.runMain(MainHelper.java:161)
[INFO]  at
org_scala_tools_maven_executions.MainWithArgsInFile.main(MainWithArgsInFile.java:26)
[INFO]
[I

**Comment by hamnis** [comment]
you cache is broken. not sure which file is broken

**Comment by dlyubimov** [comment]
which cache? maven? i've reset it several times already. and on several machines. this is very reproducible in my case (although i keep using the same set of versions of stuff)

**Comment by dlyubimov** [comment]
yeah, so the best i can figure is that one of the jars on classpath is
broken, as it is explained here
https://groups.google.com/forum/#!msg/scala-user/A-aejw44XGs/QomJ5_Chd0sJ .
however, all jars are coming from maven repo, and i reset everything
several times  -- dropped maven .m2, did `git clean -fd`, even re-installed
maven and scala and java a couple of times. None of those seem to be the
factor.

**Comment by pferrel** [comment]
I merged this with another PR I was about to push and it build fine but none of the Mahout stuff seems to run on a cluster. The build completes without errors or test failures and the cluster is working properly with the Spark shell and their examples. @hamnis have you been able to get Mahout running?

Looks like we are not doing some context creation magic correctly I'd guess. 

Starting the shell gives this error (spark-itemsimilairty error follows):

    Maclaurin:spark pat$ mahout spark-shell
    MAHOUT_LOCAL is set, so we don't add HADOOP_CONF_DIR to classpath.
    SLF4J: Class path contains multiple SLF4J bindings.
    SLF4J: Found binding in [jar:file:/Users/pat/mahout/spark/target/mahout-spark_2.10-1.0-SNAPSHOT-j    ob.jar!/org/slf4j/impl/StaticLoggerBinder.class]
    SLF4J: Found binding in [jar:file:/Users/pat/spark/assembly/target/scala-2.10/spark-assembly-1.2.1-hadoop1.2.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
    SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
    SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
    error: 
         while compiling: <init>
            during phase: typer
         library version: version 2.10.4
        compiler version: version 2.10.0
      reconstructed args: 

      last tree to typer: Literal(Constant(()))
              symbol: null
       symbol definition: null
                     tpe: Unit
           symbol owners: 
          context owners: package <empty>

    == Enclosing template or block ==

    Block( // tree.tpe=Unit
      {}
      ()
    )

    == Expanded type of tree ==

    TypeRef(TypeSymbol(final abstract class Unit extends AnyVal))

    uncaught exception during compilation: java.lang.NoSuchMethodError

    Failed to initialize compiler: NoSuchMethodError.
    This is most often remedied by a full clean and recompile.
    Otherwise, your classpath may continue bytecode compiled by
    different and incompatible versions of scala.

    java.lang.NoSuchMethodError: scala.reflect.internal.TreeInfo.firstArgument(Lscala/reflect/internal/Trees$Tree;)Lscala/reflect/internal/Trees$Tree;
    	at scala.tools.nsc.typechecker.Typers$Typer.parentTypes(Typers.scala:1550)
    	at scala.tools.nsc.typechecker.Namers$Namer.templateSig(Namers.scala:861)
    	at scala.tools.nsc.typechecker.Namers$Namer.classSig(Namers.scala:907)
    

**spark-itemsimilarity error**

    15/02/24 10:17:57 INFO spark.SecurityManager: Changing view acls to: pat,
    15/02/24 10:17:57 INFO spark.SecurityManager: Changing modify acls to: pat,
    15/02/24 10:17:57 INFO spark.SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users with view permissions: Set(pat, ); users with modify permissions: Set(pat, )
    Exception in thread "main" com.typesafe.config.ConfigException$Missing: No configuration setting found for key 'akka.event-handlers'
    	at com.typesafe.config.impl.SimpleConfig.findKey(SimpleConfig.java:115)
    	at com.typesafe.config.impl.SimpleConfig.find(SimpleConfig.java:136)
    	at com.typesafe.config.impl.SimpleConfig.find(SimpleConfig.java:142)
	at com.typesafe.config.impl.SimpleConfig.find(SimpleConfig.java:150)
    	at com.typesafe.config.impl.SimpleConfig.find(SimpleConfig.java:155)
    	at com.typesafe.config.impl.SimpleConfig.getList(SimpleConfig.java:203)
    	at com.typesafe.config.impl.SimpleConfig.getHomogeneousUnwrappedList(SimpleConfig.java:260)
    	at com.typesafe.config.impl.SimpleConfig.getStringList(SimpleConfig.java:318)
    	at akka.actor.ActorSystem$Settings.<init>(ActorSystem.scala:150)
    	at akka.actor.ActorSystemImpl.<init>(ActorSystem.scala:470)
    	at akka.actor.ActorSystem$.apply(ActorSystem.scala:111)
    	at akka.actor.ActorSystem$.apply(ActorSystem.scala:104)
    	at org.apache.spark.util.AkkaUtils$.org$apache$spark$util$AkkaUtils$$doCreateActorSystem(AkkaUtils.scala:121)
    	at org.apache.spark.util.AkkaUtils$$anonfun$1.apply(AkkaUtils.scala:54)
    	at org.apache.spark.util.AkkaUtils$$anonfun$1.apply(AkkaUtils.scala:53)
	at org.apache.spark.util.Utils$$anonfun$startServiceOnPort$1.apply$mcVI$sp(Utils.scala:1446)
    	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:141)
    	at org.apache.spark.util.Utils$.startServiceOnPort(Utils.scala:1442)
    	at org.apache.spark.util.AkkaUtils$.createActorSystem(AkkaUtils.scala:56)
    	at org.apache.spark.SparkEnv$.create(SparkEnv.scala:153)
    	at org.apache.spark.SparkContext.<init>(SparkContext.scala:203)
    	at org.apache.mahout.sparkbindings.package$.mahoutSparkContext(package.scala:95)
    	at org.apache.mahout.drivers.MahoutSparkDriver.start(MahoutSparkDriver.scala:81)
    	at org.apache.mahout.drivers.ItemSimilarityDriver$.start(ItemSimilarityDriver.scala:118)

**Comment by pferrel** [comment]
This seems to be a bug in Spark 1.2.1 SPARK-6069

Work around is to add the following either to your SparkConf in your app or -D:spark.executor.extraClassPath=/Users/pat/mahout/spark/target/mahout-spark_2.10-1.0-SNAPSHOT-dependency-reduced.jar

To the mahout spark-xyz driver, where the jar contains any class that needs to be deserialized and the path exists on all workers.

Therefor it currently looks like Spark 1.2.1 is not worth supporting.
